src/enzyme_ad/jax/Passes/LLVMToMemrefAccess.cpp

```diff
@@ -0,0 +1,159 @@
+#include "Passes.h"
+
+#include "mlir/Analysis/TopologicalSortUtils.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/MLIRContext.h"
+#include "src/enzyme_ad/jax/Dialect/Ops.h"
+
+#define DEBUG_TYPE "llvm-to-memref-access"
+
+namespace mlir {
+namespace enzyme {
+#define GEN_PASS_DEF_LLVMTOMEMREFACCESSPASS
+#include "src/enzyme_ad/jax/Passes/Passes.h.inc"
+} // namespace enzyme
+} // namespace mlir
+
+using namespace mlir;
+
+using PtrVal = TypedValue<LLVM::LLVMPointerType>;
+
+namespace mlir {
+MemRefType recoverMemRefTypeFromKernelCall(enzymexla::KernelCallOp kernelCallOp,
+                                           unsigned argIndex) {
+  auto tensorType =
+      kernelCallOp.getInputs()[argIndex].getType().dyn_cast<RankedTensorType>();
+  if (!tensorType) {
+    return nullptr;
+  }
+  // Convert tensor type to memref type
+  return MemRefType::get(tensorType.getShape(), tensorType.getElementType());
+}
+
+} // namespace mlir
+
+struct LLVMToMemrefAccessPass
+    : public enzyme::impl::LLVMToMemrefAccessPassBase<LLVMToMemrefAccessPass> {
+  using LLVMToMemrefAccessPassBase::LLVMToMemrefAccessPassBase;
+
+  void runOnOperation() override {
+    Operation *op = getOperation();
+
+    SymbolTableCollection symbolTableCollection;
+    SymbolTable symbolTable = symbolTableCollection.getSymbolTable(op);
+
+    MLIRContext *context = &getContext();
+
+    // Find all enzymexla.kernel_call ops and recover memref types for callees
+    op->walk([&](enzymexla::KernelCallOp kernelCallOp) {
+      auto *symbolOp = symbolTableCollection.lookupNearestSymbolFrom(
+          kernelCallOp, kernelCallOp.getFnAttr());
+      auto fn = cast<FunctionOpInterface>(symbolOp);
+
+      // 1. Recover memref types for all arguments of the callee
```

Comment:
@ivanradanov and I discovered an issue last night (fixed in https://github.com/EnzymeAD/Enzyme-JAX/pull/303/files#diff-226e98da706121c678606489ff574cfe82f6143f1cfeb844c99d4d7dc2a94deb) where you aren't allowed to have non-llvm-typed arguments in llvm.func's, so we had to raise to a func.func iirc

Reply:
Yes we are not allowed to have memref arguments in llvm.func, that's why here newFunc is func::FuncOp instead.